src/coreclr/nativeaot/Runtime/stdcppshim.cpp

```diff
@@ -0,0 +1,30 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include "CommonTypes.h"
+
+void __cxa_pure_virtual() 
```

Comment:
`__cxa_pure_virtual` (which indicates bug or unusual interaction) is typically implemented as while(true) loop or abort(), so logically we won't miss much if our implementation is preferred over that of libstdc++.

For static linkage, if there are multiple conflicting symbols  we can use `--allow-multiple-definition` (as we anyway need today to link multiple NAOT static objects: https://github.com/dotnet/runtime/issues/70277#issuecomment-1157574381).

Reply:
I am worried about somebody statically linking NAOT static object and static C++ libraries together.

I am not worried about somebody linking multiple NAOT static objects together. I do not think that it is interesting scenario.

Reply:
Would weak symbols be a reasonable solution here? If we make this a weak symbol, then it will only be used when no C++ libraries are statically linked with a NAOT static object.

Reply:
> I am worried about somebody statically linking NAOT static object and static C++ libraries together.

If two objects are providing the same symbol, then `--allow-multiple-definition` avoids the `duplicate symbols` warning.

Reply:
What are the C++ constructs that end up producing `__cxa_pure_virtual` calls? Can we get rid of them instead?

Reply:
Both compilers (clang++ and g++) expect `__cxa_pure_virtual` symbol in the binary when we use a pure virtual function `T M() = 0;`, regardless the code violates the rule of calling PVF from ctor or dtor (which we don't).

This is what we get:
 
```
 crossgen2 -> /runtime/artifacts/bin/coreclr/Linux.x64.Debug/crossgen2/linux-x64/crossgen2.dll
  Generating native code
  /usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x25
  /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x10): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x18): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x20): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x28): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x30): undefined reference to `__cxa_pure_virtual'
  /usr/bin/ld: /runtime/artifacts/bin/coreclr/Linux.x64.Debug/aotsdk/libRuntime.ServerGC.a(objecthandle.cpp.o):(.data.rel.ro._ZTV14IGCHandleStore[_ZTV14IGCHandleStore]+0x38): more undefined references to `__cxa_pure_virtual' follow
```

The demangled form of `_ZTV14IGCHandleStore` is `vtable for IGCHandleStore`.

While googling about the compiler internals, I stumbled upon https://elegantinvention.com/blog/information/smaller-binary-size-with-c-on-baremetal-g/. The author was also trying to build with -nostdlib and ran into this issue due to the precense of pure virtual in their code base.

> Can we get rid of them instead?

`IGCHandleStore` is defined in the gcinterface, which is now a public contract (by means of standalone-gc's 3P implementations). Otherwise, I was going with replacing all occurrences of `IGCHandleStore` with `GCHandleStore` and deleting the class. :)

Reply:
We do not support 3P gc implementations. GC interface is not a public contract. We only care about the GC contract to the point of being able swap the GC implementations for some range of builds.

What is special about IGCHandleStore? We use similar pattern for other types too (e.g. `IGCHeap`) and they are not causing any problems.

Can we define `#define ABSTRACT_IMPL` macro and use it instead of `= 0;` for abstract types? It should get rid of dependency on `__cxa_pure_virtual`.

Reply:
> We do not support 3P gc implementations. GC interface is not a public contract.

Cool. I found implementations in the wild inheriting from IGCHandleManager to provide their own manager in custom standalone gc. Good to know we don't consider it as a public contract.

> What is special about IGCHandleStore? We use similar pattern for other types too (e.g. IGCHeap) and they are not causing any problems.

Nothing other than `g_gcGlobalHandleStore` is first one used in the initialization?

I still don't understand what is wrong with `__cxa_pure_virtual`. Could you give an example what scenario will break? There is no difference in static linking commands which I showed here https://github.com/dotnet/runtime/issues/70277.

Reply:
`--allow-redefinition` can hide serious problems, it looks very fragile. I do not want to be telling anybody to use `--allow-redefinition` because of then we will need to be spending time debugging the inscrutable problems caused by it.